### PR TITLE
8282470: Eliminate useless sign extension before some subword integer operations

### PR DESCRIPTION
**Pull request by fg1417**
Some loop cases of subword types, including byte and short, can't be vectorized by C2's SLP. Here is an example:
```
short[] addShort(short[] a, short[] b, short[] c) {
  for (int i = 0; i < SIZE; i++) {
    b[i] = (short) (a[i] + 8); //  line A
    sres[i] = (short) (b[i] + c[i]); // line B
  }
}
```
However, similar cases of int/float/double/long/char type can be vectorized successfully.

The reason why SLP can't vectorize the short case above is that, as illustrated here[1], the result of the scalar add operation on *line A* has been promoted to int type. It needs to be narrowed to short type first before it can work as one of source operands of addition on *line B*. The demotion is done by left-shifting 16 bits then right-shifting 16 bits. The ideal graph for the process is showed like below.
image

In SLP, for most short-type cases, we can determine the precise type of the scalar int-type operation and finally execute it with short-type vector operations[2], except rshift opcode and abs in some situations[3]. But in this case, the source operand of RShiftI is from LShiftI rather than from any LoadS[4], so we can't determine its real type and conservatively assign it with int type rather than real short type. The int-type opearation RShiftI here can't be vectorized together with other short-type operations, like AddI(line B). The reason for byte loop cases is the same. Similar loop cases of char type could be vectorized because its demotion from int to char is done by `and` with mask rather than `lshift_rshift`.

Therefore, we try to remove the patterns like `RShiftI _ (LShiftI _ valIn1 conIL ) conIR` in the byte/short cases, to vectorize more scenarios. Optimizing it in the mid-end by i-GVN is more reasonable.

What we do in the mid-end is eliminating the sign extension before some subword integer operations like:

```
int x, y;
short s = (short) (((x << Imm) >> Imm) OP y); // Imm <= 16
```
to
```
short s = (short) (x OP y);
```

In the patch, assuming that `x` can be any int number, we need guarantee that the optimization doesn't have any impact on result. Not all arithmetic logic OPs meet the requirements. For example, assuming that `Imm` equals `16`, `x` equals `131068`,
`y` equals `50` and `OP` is division`/`, `short s = (short) (((131068 << 16) >> 16) / 50)` is not equal to `short s = (short) (131068 / 50)`. When OP is division, we may get different result with or without demotion before OP, because the upper 16 bits of division may have influence on the lower 16 bits of result, which can't be optimized. All optimizable opcodes are listed in StoreNode::no_need_sign_extension(), whose upper 16 bits of src operands don't influence the lower 16 bits of result for short
type and upper 24 bits of src operand don't influence the lower 8 bits of dst operand for byte.

After the patch, the short loop case above can be vectorized as:
```
movi    v18.8h, #0x8
...
ldr     q16, [x14, #32] // vector load a[i]
// vector add, a[i] + 8, no promotion or demotion
add     v17.8h, v16.8h, v18.8h
str     q17, [x6, #32]  // vector store a[i] + 8, b[i]
ldr     q17, [x0, #32]  // vector load c[i]
// vector add, a[i] + c[i], no promotion or demotion
add     v16.8h, v17.8h, v16.8h
// vector add, a[i] + c[i] + 8, no promotion or demotion
add     v16.8h, v16.8h, v18.8h
str     q16, [x11, #32]  //vector store sres[i]
...
```

The patch works for byte cases as well.

Here is the performance data for micro-benchmark before and after this patch on both AArch64 and x64 machines. We can observe about ~83% improvement with this patch.

on AArch64:
Before the patch:
Benchmark (length)  Mode  Cnt    Score    Error  Units
addB       523      avgt   15  401.521 ±  0.033  ns/op
addS       523      avgt   15  401.512 ±  0.021  ns/op

After the patch:
Benchmark (length)  Mode  Cnt    Score   Error  Units
addB       523      avgt   15   68.444 ± 0.318  ns/op
addS       523      avgt   15   69.847 ± 0.043  ns/op

on x86:
Before the patch:
Benchmark (length)  Mode  Cnt    Score    Error  Units
addB       523      avgt   15  454.102 ± 36.180  ns/op
addS       523      avgt   15  432.245 ± 22.640  ns/op

After the patch:
Benchmark (length)  Mode  Cnt    Score    Error  Units
addB       523      avgt   15   75.812 ±  5.063  ns/op
addS       523      avgt   15   72.839 ± 10.109  ns/op

[1]: https://github.com/openjdk/jdk/blob/6013d09e82693a1c07cf0bf6daffd95114b3cbfa/src/hotspot/share/opto/superword.cpp#L3241
[2]: https://github.com/openjdk/jdk/blob/6013d09e82693a1c07cf0bf6daffd95114b3cbfa/src/hotspot/share/opto/superword.cpp#L3206
[3]: https://github.com/openjdk/jdk/blob/6013d09e82693a1c07cf0bf6daffd95114b3cbfa/src/hotspot/share/opto/superword.cpp#L3249
[4]: https://github.com/openjdk/jdk/blob/6013d09e82693a1c07cf0bf6daffd95114b3cbfa/src/hotspot/share/opto/superword.cpp#L3251

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.java.net/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8282470](https://bugs.openjdk.java.net/browse/JDK-8282470): Eliminate useless sign extension before some subword integer operations

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7954/head:pull/7954` \
`$ git checkout pull/7954`

Update a local copy of the PR: \
`$ git checkout pull/7954` \
`$ git pull https://git.openjdk.java.net/jdk pull/7954/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7954`

View PR using the GUI difftool: \
`$ git pr show -t 7954`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7954.diff">https://git.openjdk.java.net/jdk/pull/7954.diff</a>

</details>

### CODE REVIEW

**Comment by vnkozlov** [comment]
@fg1417  Thank you for proposing these changes. I hope you still want to proceed with it.

**Comment by fg1417** [comment]
> @fg1417 Thank you for proposing these changes. I hope you still want to proceed with it.

I updated the patch to latest JDK. Could you please help review it, @vnkozlov ? Thanks.

**Review by vnkozlov — COMMENTED** [review]
Looks reasonable. Let me test it.

**Comment by vnkozlov** [comment]
Note, these changes need second review (it is not trivial).

**Comment by fg1417** [comment]
Thanks a lot for your kind review and  also for your test work, @vnkozlov .

**Comment by DamonFool** [comment]
This patch can auto-vectorize `testShort1`, but fail to do so with `testShort2`.
```cpp
    public static void testShort1(short[] a, short[] b, short[] c, short[] d) {
        for (int i = 0; i < a.length; i++) {
            b[i] = (short)(a[i] + 1);
            d[i] = (short)(b[i] + c[i]);
        }
    }

    public static void testShort2(short[] a, short[] b, short[] c, short[] d) {
        for (int i = 0; i < a.length; i++) {
            b[i] = (short)(a[i] + 1);
            d[i] = (short)(b[i] + c[i] + 1);
        }
    }
```

So may I ask if it's possible to also auto-vectorize `testShort2` ?
Thanks.

**Comment by DamonFool** [comment]
So the current implementation only works for limited scenarios, right?

I'm not sure if there is a good way to eliminate the useless sign extension in `testShort2`.
But I really hope this opt can be used for more situations.

Let's see if someone has a good idea.
Thanks.

File: test/micro/org/openjdk/bench/vm/compiler/TypeVectorOperations.java [hunk 1]
```diff
@@ -70,7 +75,8 @@ public void init() {
         resF = new float[COUNT];
 
         for (int i = 0; i < COUNT; i++) {
-            shorts[i] = (short) r.nextInt(Short.MAX_VALUE + 1);
+            shortA[i] = (short) r.nextInt(Short.MAX_VALUE + 1);
```
Comment (JohnTortugo) [inline test/micro/org/openjdk/bench/vm/compiler/TypeVectorOperations.java:78]: Why is there a "+1" here?

File: test/hotspot/jtreg/compiler/c2/irTests/TestIRStoreCorBAddIMask.java [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+    private byte BYTE_OUT = 0;
+    private char CHAR_OUT = 0;
+    private short SHORT_OUT = 0;
+
+    public static void main(String[] args) {
+        TestFramework.run();
+    }
+
+    @Test
+    @Arguments({Argument.DEFAULT, Argument.DEFAULT})
```
Comment (JohnTortugo) [inline test/hotspot/jtreg/compiler/c2/irTests/TestIRStoreCorBAddIMask.java:57]: NIT: Won't be better to use RANDOM_EACH in the parameters of this and other methods?

File: src/hotspot/share/opto/memnode.hpp [hunk 3]
```diff
@@ -559,8 +559,12 @@ class StoreNode : public MemNode {
   virtual bool cmp( const Node &n ) const;
   virtual bool depends_only_on_test() const { return false; }
 
-  Node *Ideal_masked_input       (PhaseGVN *phase, uint mask);
-  Node *Ideal_sign_extended_input(PhaseGVN *phase, int  num_bits);
+  bool used_only_for_this_opcode(Node* in);
```
Comment (JohnTortugo) [inline src/hotspot/share/opto/memnode.hpp:562]: NIT: can these methods be const and/or static?

File: src/hotspot/share/opto/memnode.hpp [hunk 4]
```diff
@@ -559,8 +559,12 @@ class StoreNode : public MemNode {
...
+  bool no_need_sign_extension(int opc);
+  bool is_masked_input(Node* input, PhaseGVN* phase, uint mask);
+  bool is_sign_extended_input(Node* input, PhaseGVN* phase, int num_bits);
+
+  Node* Ideal_masked_or_sign_extended_input(PhaseGVN* phase, uint mask, int num_bits);
```
Comment (JohnTortugo) [inline src/hotspot/share/opto/memnode.hpp:567]: NIT: Looks like this method name follows the coding style.

**Comment by rose00** [comment]
I will believe more in this transformation if it covers a wide range of expressions, not just a few specially-matched patterns `r = (short)(x + (int)y)`.

Searching to an unbounded depth in one transform is usually a sign something is formulated wrong.  Often it's possible to break the search up into a series of incremental transformations that, taken together, get to the desired end.

In this case, I guess you want to be able to transform `r = (short)XYZ` such that all needless short-to-int conversions inside of `XYZ` are removed.  A short-to-int conversion `S2I(x)` is needless if (a) it is an operand of a carryless or leftward-carrying ALU operation `S2I(x) op y`, and (b) all consumers of the operation only care about the low 16 bits of the result.

I think that, whether or not this patch is accepted, and certainly before the next patch which handles ternary adds or SubI or MulI in a similar manner is proposed, we should tackle that more general problem.  (And not just for 16-bit ints.  If I do a mask `(x + y) & 0x7F` or sign extension `(x + y) <<7 >>7` I should be able to benefit from the same kinds of reasoning.

**Comment by fg1417** [comment]
Thanks for your review and kind suggestions, @rose00 @DamonFool @JohnTortugo 

I really understand your concern and agree. Frankly, I attempted to seek for a common solution to cover a wider range but failed. But I will give a more try to figure it out if we can make this transformation not that specific and more general, before continuing proceeding with this patch. Thanks.

**Pull request closed without merging**